pkg/didcomm/common/messenger/messenger.go

```diff
@@ -0,0 +1,36 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package messenger
+
+import "github.com/hyperledger/aries-framework-go/pkg/didcomm/common/service"
+
+// The messenger package is responsible for the handling of communication between agents.
+// It takes care of threading, timing, etc.
+// Each message that we are going to send should be DIDCommMsg type.
+// e.g we have type model.Ack{Type: "type", ID: "id"}
+// it should be converted to: map[@id:id @type:type]
+// NOTE: The package modifies message data by JSON tag name according to aries-rfcs.
+//       Fields like @id, ~thread , etc. are redundant and may be rewritten.
+
+// Client provides methods for the communication
+type Client interface {
+	// HandleMessage receives incoming message and prepares data needed for the communication according to the options.
+	// Options should be provided by the service.
+	HandleMessage(msg service.DIDCommMsg, myDID, theirDID string, opts ...interface{})
+
+	// ReplyTo replies to the message by given msgID.
+	// Keeps threadID in the *decorator.Thread.
+	// Using this function means that communication will be on the same thread.
+	ReplyTo(msgID string, msg service.DIDCommMsg)
+
+	// Send sends the message by starting a new thread.
+	Send(msg service.DIDCommMsg, myDID, theirDID string)
+
+	// ReplyToNested sends the message by starting a new thread.
+	// Keeps parent threadID in the *decorator.Thread
+	ReplyToNested(msgID string, msg service.DIDCommMsg)
```

Comment:
What is the use case for this?

Nested threads typically start a new protocol. In those situations, this method's signature implies that a protocol service will know how to build messages of another protocol.

Please clarify.

Reply:
Correct, nested starts a new protocol. But also it requires parent thread id. Instead of threadID we are using messageID. So, no matter which service starts a new protocol(nested) we should pass the messageID which causes this action. Then, we will get the correct thread ID by messageID .

Reply:
@soluchok my issue with this is that protocol services should not know the details of the messages of other protocols.

Reply:
ok, then I will change it to threadID

Reply:
Done!